### PR TITLE
Fix bottom display when pattern is near the edge

### LINKED ISSUE #733: [BUG] Lower display is blank with edges wider than image
AYAB software version: 1.0 Beta3
Computer/OS: Mac
Knitting machine: 930
AYAB hardware: shield

If you set start and stop edges that are wider than the image, and the stop edge is anywhere to the left of needle Right2, it can get into a situation where it is not displaying anything in the lower display, and not updating the upper display, though it is knitting correctly.

1. Start with all settings at default
2. Open test image uc3_3-x15.png
3. Set the edges. to L34 - R2, so the 30st image is centered in 36 stitches
4. Click Simulate, or Knit.
5. If you actually (air)knit it, it does knit in the right place
6. Wait for Simulate to finish, it is actually doing something, you get the ending sounds.

Expected: It should step through and display, with the image centered within its area and 3 squares of white on each side. 
Actual: the lower display is blank, and the upper display doesn't step up. The needle numbers displayed in the lower box are from the last time it worked. It does knit in the right place.

### PR DESCRIPTION
**Pull request by jonathanperret**
## Problem

Fixes #733.

## Proposed solution

Capture d’écran 2024-12-04 à 23 41 24

The bug was caused by an edge case where the `midline` value was apparently wrong but unfortunately I could not really make sense of what that value was supposed to represent. Instead I changed the code to use the index of the first patterning needle, along with the machine width, to compute the displayed (orange/green) needle numbers.

Along the way I simplified the logic to transmit the current status to the UI. Some values were transported as extra arguments alongside the `Status` object; moving them into the `Status` class made the code more regular. I also removed a couple of parameters that were entirely unused.

I also fixed what appeared to be a logic error that caused `Pass 1` to be shown in row headers in single-bed mode (where each row always has a single pass), and never be shown on ribber modes (where multiple passes are required per pattern row) — interestingly this appears to be a regression of a once-fixed issue: https://github.com/AllYarnsAreBeautiful/ayab-desktop/issues/548. Now single-bed knitting shows no pass number, and multi-pass modes show `Pass x/n`.

For code review, a note on the changes in `knitprogress.py`: most of the lines that are modified have no change in logic, but the configured style checker (`flake8`), which apparently had never been run on that file, required splitting a number of lines before I could commit the file. Sorry for the churn.

## How to test

Here's a test release built from the code in this PR: https://github.com/jonathanperret/ayab-desktop/releases/tag/1.0.0-knitprogress-1

## Release Notes

- **New Features**
  - Introduced new attributes in the status management system: `machine_width`, `knit_start_needle`, and `passes_per_row`.

- **Improvements**
  - Streamlined the `Control`, `KnitProgress`, and `SignalSender` classes for better status management and code readability.
  - Simplified signal emission in the `SignalReceiver` class to enhance clarity.

- **Bug Fixes**
  - Removed unnecessary parameters in various methods to ensure smoother operation and reduce complexity.

### CODE REVIEW

File: src/main/python/main/ayab/knitprogress.py [hunk 1]
```diff
@@ -236,9 +264,14 @@ def onStitchSelect(self, current: QTableWidgetItem | None) -> None:
         if current is None:
             self.__progbar.set_selection_label("")
             return
-        if self.horizontalHeaderItem(current.column()).foreground().color().red() == 187:
+        if (
+            self.horizontalHeaderItem(current.column()).foreground().color().red()
+            == 187
+        ):
```
[the remark this answers is not in the document]
Comment (jonathanperret) [reply src/main/python/main/ayab/knitprogress.py:270]: I don't like it either but it has nothing to do with this PR. And I don't think naming the constant fundamentally improves the logic here anyway.

**Comment by dl1com** [comment]
As this fixes not only #733 but also tweaks some more things, I'd love to get some feedback of testing.

**Comment by jonathanperret** [comment]
> As this fixes not only #733 but also tweaks some more things, I'd love to get some feedback of testing.

Agreed, more feedback is always welcome. @Adrienne200 did you get a chance to try this one out maybe?

**Comment by jonathanperret** [comment]
Reported working on Discord by @Adrienne200 .

**Review by dl1com — APPROVED** [review]

**Pull request merged**